### PR TITLE
GJULE - new GlassFish logging backend

### PR DESCRIPTION
**Pull request by dmatej**
### Targets

- Fix long-standing issues in GlassFish logging system
- Make logging of startup+shutdown configurable and even possible
  - previous impl lost all logs more verbose than INFO logged before the logging was configured from logging.properties
  - avoid usage of earlyLogs to have at least some logs
- Keep JUL standards - main source of the configuration is logging.properties file, reconfigurations have lifecycle and must ensure consistency
- Better performance
- Cover logging by tests (more tests to come in near future)
- Separate the logging system between the JVM feature extension and HK2 services which **use** and configure the logging system, read logs, initiate rolling the server.log file, etc.

### Changes

- server.log now by default prints microseconds instead of millis
- default timestamp format is now compliant with ISO 8601
- GF ClassLoaders have toString - makes reading logs easier
- Fixed FelixPrettyPrinter usage when server doesn't start because of some dependency
- bootstrap libraries - used as bootstrap dependencies, layer between JDK and OSGi.
- AccessLog now uses Java Time API instead of SDF and util.Date
  - GF doesn't have any tests for that, tested manually
  - now much faster, however still uses same obsoleted format
- `SysLogHandler` configuration enhanced and tested manually on Kubuntu 22.04.1
- `ConsoleLogHandler` replaced by `SimpleLogHandler` and now uses STDERR for WARNINGS+ and STDOUT for INFO- as it is usual in Linux systems.
- new `OneLineLogFormatter` - for easily readable logs without noise; especially useful for unit tests.
- WALL_CLOCK_START JVM option set on startup - was already hidden in the code, it is pretty useful when you list java programs on system where is running many JVM/GlassFish instances and you need to find quickly which one is the one; also easy to see how long it is running or to find related events.

### Following pull requests

- Documentation
- Integration tests (clusters, reconfigurations, restarts, access logs, virtual server logs)
- More fixes (suspicious code in GF, see FIXMEs, explain+document+cover by tests OR fix/remove)

### CODE REVIEW

**Review by OndroMih — CHANGES_REQUESTED** [review]
I would add some caching to the new toString methods to improve performance.

File: nucleus/admin/cli/src/main/java/com/sun/enterprise/admin/cli/DirectoryClassLoader.java [hunk 1]
```diff
@@ -16,72 +17,88 @@
...
-        });
-        if (fjars == null)
-            throw new IOException(strings.get("DirError", dir));
-        URL[] jars = new URL[fjars.length];
-        for (int i = 0; i < fjars.length; i++)
-            jars[i] = fjars[i].toURI().toURL();
-        return jars;
+
+    @Override
+    public String toString() {
```
[the remark this answers is not in the document]
Comment (dmatej) [reply nucleus/admin/cli/src/main/java/com/sun/enterprise/admin/cli/DirectoryClassLoader.java:97]: Actually it is logged just when some error occurs, but even then much more expensive is printing a stacktrace. 
Originally there was a method printing list of urls maintained by one classloader. But where I needed it, it printed just CL class name and hashcode (class/resource not found, but by which CL?). This is why I implemented toString here.

**Comment by dmatej** [comment]
That test which failed on Jenkins passed locally in 7 seconds.

**Review by arjantijms — APPROVED** [review]

File: nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/GlassFishLogManagerLifeCycleTest.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        reconfigureToBlockingHandler();
+        setResolveLevelWithIncompleteConfiguration(false);
+        assertEquals(GlassFishLoggingStatus.CONFIGURING, MANAGER.getLoggingStatus(), "after reconfigureToBlockingHandler");
+
+        doLog(Level.FINEST, "message0");
+        doLog(Level.INFO, "message1");
+        MANAGER.reconfigure(originalCfg, () -> LOG.setLevel(Level.FINEST), null);
+        assertFalse(isResolveLevelWithIncompleteConfiguration(), "isResolveLevelWithIncompleteConfiguration");
+
+        final List<GlassFishLogRecord> logRecords = COLLECTOR.getAll();
```
Comment (pzygielo) [inline nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/GlassFishLogManagerLifeCycleTest.java:213]: Have seen recently:
```
2025-01-26T08:21:39.9703043Z Both records must pass via startup queue (4 failures)
2025-01-26T08:21:39.9704392Z >--java.lang.AssertionError: record 0
2025-01-26T08:21:39.9704815Z Expected: <FINEST>
2025-01-26T08:21:39.9705101Z      but: was <INFO>
2025-01-26T08:21:39.9705451Z >--java.lang.AssertionError: record 0
2025-01-26T08:21:39.9705854Z Expected: "message0"
2025-01-26T08:21:39.9706155Z      but: was "message1"
2025-01-26T08:21:39.9706532Z >--java.lang.AssertionError: record 1
2025-01-26T08:21:39.9706915Z Expected: <INFO>
2025-01-26T08:21:39.9707189Z      but: was <FINEST>
2025-01-26T08:21:39.9707521Z >--java.lang.AssertionError: record 1
2025-01-26T08:21:39.9707903Z Expected: "message1"
2025-01-26T08:21:39.9708195Z      but: was "message0"
```

Timing issue?

Looks like records are collected in order different than expected by assertions.
From the message _Both records must pass via startup queue_ I'd not suspect the order to be crucial. I believe that implementation is rather fine here and it's test that should follow.

**Pull request merged**